options(stringsAsFactors=FALSE)
setwd("/m/jdrfdn_scratch/users/ccr5ju/AA_Immuno/hla_assoc")
#install.packages("lmtest", repos="https://cloud.r-project.org")
library(lmtest)
library(rtf)

genofile = "derived_data/genotypes_recode.raw.4d"
rtffilename = "tables/table1b.doc"
phenofile ="../data_original/pc2.txt"

### Get best guess genotypes
#args=commandArgs(trailingOnly=TRUE)
#if (length(args)==3) {
#	genofile = args[1]
#	rtffilename = args[2]
#	phenofile = args[3]
#} else {
#	stop("Incorrect number of arguments specified\n")
#}

rtffile <- RTF(rtffilename)  # this can be an .rtf or a .doc
geno = read.table(genofile, comment.char="~", header=TRUE)
pheno = read.table(phenofile, comment.char="~", header=TRUE)
covariates = names(pheno)[c(-1,-2)]
df = merge(geno, pheno, by="IID")
df$t1d = df$PHENOTYPE-1
genes = c("HLA_A", "HLA_B", "HLA_C", "HLA_DRB1", "HLA_DQA1", "HLA_DQB1", "HLA_DPA1", "HLA_DPB1")

outAll = list()

for (k in 1:length(genes)) {
	gene = genes[k]
	alleles=names(geno)[grep(gene, names(geno))]
	
	#Quality filtering (exclude subjects whose alleles do not add to 2 for a given locus)
	X = geno[,alleles]; rownames(X) <- geno$IID
	exclude = rownames(X)[!rowSums(X)==2]
	df_filt = df[!df$IID%in%exclude,]
	
	#Collapse rare alleles
	var_count = NULL 
	for (i in 1:length(alleles)) {
		var_count[i] = sum(df_filt[,alleles[i]])
	}
	common_alleles = alleles[var_count>=30]
	rare_alleles = alleles[var_count<30]
	if (length(rare_alleles)==0) {
		alleles_new = common_alleles		
	} else if (length(rare_alleles)==1) {
		alleles_new = c(common_alleles, rare_alleles)
	} else if (length(rare_alleles)>1) {		
		groups = paste(unlist(lapply(strsplit(rare_alleles, split="_"),"[[", 1)), 
				unlist(lapply(strsplit(rare_alleles, split="_"),"[[", 2)),
				substring(unlist(lapply(strsplit(rare_alleles, split="_"),"[[", 3)), 1, 2),
				unlist(lapply(strsplit(rare_alleles, split="_"),"[[", 4)), sep="_")
		rare_allele_dict = list()
		common_groups = NULL
		for (i in 1:length(unique(groups))) {
			group = unique(groups)[i]
			alleles_in_group = rare_alleles[groups==group]
			rare_allele_dict[[i]] = list(group=group, alleles=alleles_in_group, count=sum(df_filt[,alleles_in_group]))
			if (length(alleles_in_group)>1) {
				df_filt[,group] = rowSums(df_filt[,alleles_in_group])
				if (sum(df_filt[,group]) >= 30) {
					common_groups = c(common_groups, group)	
				}
			}		
		}
		alleles_new = c(common_alleles, common_groups) 
	}
	common_df = data.frame(df_filt[,c("t1d", alleles_new, "PC1", "PC2")])
		

	out.glm.null = glm(as.formula(paste("t1d~",paste(covariates,collapse="+"))), family="binomial", data=common_df)
	roundTo = 2
	outDF = data.frame() 
	for (i in 1:length(alleles_new)) {
		
		allele = alleles_new[i]
		
		#Get allele info
		case_count = sum(common_df[common_df$t1d==1,allele])
		control_count = sum(common_df[common_df$t1d==0,allele])
		case_freq = sum(common_df[common_df$t1d==1,allele])/(2*dim(common_df[common_df$t1d==1,])[1])
		case_percent = format(round(case_freq*100, roundTo), nsmall=roundTo, trim=TRUE)
		control_freq = sum(common_df[common_df$t1d==0,allele])/(2*dim(common_df[common_df$t1d==0,])[1])
		control_percent = format(round(control_freq*100, roundTo), nsmall=roundTo, trim=TRUE)
		
		OR_unadjusted = format(round((case_freq/(1-case_freq))/(control_freq/(1-control_freq)), roundTo), nsmall=roundTo, trim=TRUE)
		
		#Fit model
		out.glm.full = glm(as.formula(paste("t1d~",paste(c(allele, covariates),collapse="+"))), family="binomial", data=common_df)
		#out.glm.full = glm(as.formula(paste("t1d~",paste(c(allele),collapse="+"))), family="binomial", data=common_df)
		full.coeff = coef(summary(out.glm.full))[allele,]
		full.OR = format(exp(full.coeff[1]), digit=roundTo)
		full.ub = exp(full.coeff[1]+1.96*full.coeff[2])
		full.lb = exp(full.coeff[1]-1.96*full.coeff[2])
		full.p = format(full.coeff[4], scientific=TRUE, digits=roundTo)
		lrt_pvalue = format(lrtest(out.glm.full, out.glm.null)$"Pr(>Chisq)"[2], scientific=TRUE, digits=roundTo)
		
		outDF = rbind(outDF, c(allele, case_percent, control_percent, OR_unadjusted, full.OR, full.p, lrt_pvalue)) 
			
	}
	names(outDF) <- c("allele", "Case", "Control", "OR_unadj", "OR", "P", "p_lrt")
	outDF$Gene=unlist(lapply(strsplit(outDF$allele, split="_"),"[[", 2))
	outDF$Allele=unlist(lapply(strsplit(outDF$allele, split="_"),"[[", 3))
	
	outAll[[k]] = outDF
	
	OUT.TABLE = outDF[order(outDF$Allele, decreasing=FALSE),c("Allele", "Case", "Control", "OR", "P")]
	addParagraph(rtffile, paste("Table 1-", k, " Association of T1D in African Americans with ", gene, sep=""))
	addTable(rtffile, OUT.TABLE)
	addParagraph(rtffile,"Odds ratios estimated and p-values generated by logistic regression, adjusting for 2 principal components. Only 4-digit alleles with total allele count >30 are analyzed. Remaining rare 4-digit alleles are combined into 2-digit groups. \n\n")
}

done(rtffile)


outCombined = rbind(outAll[[1]], outAll[[2]], outAll[[3]], outAll[[4]], outAll[[5]], outAll[[6]], outAll[[7]], outAll[[8]])
outCombined$P_num = as.numeric(outCombined$P)
outCombined$p_lrt_num = as.numeric(outCombined$p_lrt)
outCombined[order(outCombined$p_lrt_num),]


